### PR TITLE
Events DB: ReadMarker table

### LINKED ISSUE #764: Event db: ReadMarker table
We need an event database [ReadMarker](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/models.go#L1166) table. this is populated during [read_redeem](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/blobber.go#L344).

We need code for the new table and add ReadMarker method. There five foreign keys into other tables to arrange.

We need to emit an add ReadMarker event in[commitBlobberRead](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/blobber.go#L348) which includes the input ReadMarker sent by the blobber plus the transaction hash.
ReadSize is the difference between the last readCounter and this one [commitBlobberRead](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/blobber.go#L419) .

The [/latestreadmarker](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/handler.go#L194) can be reworked to use this table.

https://github.com/0chain/0chain/issues/789 adds new API methods to access this table.

| Event db ReadMarker | type   | ReadMarker   | type      | notes                              |
|---------------------|--------|--------------|-----------|------------------------------------|
| Client              | uint   | ClinetId     | string    | foreign key into user table        |
| BlobberId           | uint   | BlobberId    | string    | foreign key into blobber table     |
| AllocationId        | uint   | AllocationId | string    | foreign key into allocation table  |
| TransactionId       | uint   | t.Hash       | string    | foreign key into transaction table |
| OwnerId             | uint   | OwnerId      | string    | foreign key into user table        |
| Timestamp           | int64  | Timestamp    | Timestamp |                                    |
| ReadCounter         | int64  | ReadCounter  | int64     |                                    |
| ReadSize         | int64  | ReadCounter  | int64     |                                    |
| Signature           | string | Signature    | string    |                                    |
| Suspend             | int64  | Suspend      | int64     |                                    |
| PayerId             | uint   | PayerID      | string    | foreing key into user table        |
| AuthTicket          | string | AuthTicket   | JSON      |      searchable index                              |
| BlockNumber             | int64  | txn.Round      | int64     |                                    |

### LINKED ISSUE #789: Event db API: read markers from authTicket
https://github.com/0chain/0chain/issues/764 introduces the ReadMarker table. 

We want an endpoint that has an authToken as input and returns all readMarkers with that authTicket.
```sql
select * from ReadMarker where AuthTicket = params.Get("auth_ticket")
```

We also need an endpoint that will return all read markers from an alloactionId.
```sql
select * from ReadMarker where AllocationId = params.Get("allocation_id")
```

We want an endpoint to return the number of reads for an allocation.
```sql
select count(*) from ReadMarker where AllocationId = params.Get("allocation_id")
```

### PR DESCRIPTION
**Pull request by ssardana96**
Tasks to complete before merging PR:
- [x]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [x]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [x]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

### CODE REVIEW

**Comment by ssardana96** [comment]
Created issue #897 for pending foreign keys

File: code/go/0chain.net/smartcontract/dbs/event/readmarker.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	AllocationID  string  `json:"allocation_id"`
+	TransactionID string  `json:"transaction_id"`
+	OwnerID       string  `json:"owner_id"`
+	Timestamp     int64   `json:"timestamp"`
+	ReadCounter   int64   `json:"read_counter"`
+	ReadSize      float64 `json:"read_size"`
+	Signature     string  `json:"signature"`
+	PayerID       string  `json:"payer_id"`
+	AuthTicket    string  `json:"auth_ticket"`
+	BlockNumber   int64   `json:"block_number"`
```
Comment (kushthedude) [inline code/go/0chain.net/smartcontract/dbs/event/readmarker.go:23]: Dont we need to define primary key with gorm.Model?

File: code/go/0chain.net/smartcontract/storagesc/handler.go [hunk 2]
```diff
@@ -242,6 +243,75 @@ func (ssc *StorageSmartContract) LatestReadMarkerHandler(ctx context.Context,
...
+	count, err := balances.GetEventDB().CountReadMarkersFromQuery(query)
+	if err != nil {
+		return nil, common.NewErrInternal("can't count read markers", err.Error())
+	}
+
+	return struct {
+		ReadMarkersCount int64 `json:"read_markers_count"`
+	}{
+		ReadMarkersCount: count,
+	}, nil
```
Comment (kushthedude) [inline code/go/0chain.net/smartcontract/storagesc/handler.go:311]: TIL for me

**Review by kushthedude — DISMISSED** [review]

File: code/go/0chain.net/smartcontract/dbs/event/readmarker.go (hunk 1 above)
Comment (ssardana96) [reply code/go/0chain.net/smartcontract/dbs/event/readmarker.go:23]: by default ID is primary key

**Comment by ssardana96** [comment]
while integration testing found few bugs which I've added in same PR:
- change in remove duplicate logic in events (added tags and type check).
- transactions were not getting added to db (0 rows).
- transaction was taking magicBlock hash, instead of block hash.

**Review by Sriep — APPROVED** [review]
Looks ok, but we will need system tests for the two new endpoints GetReadMarkersHandler and GetReadMarkersCount.

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feature/zcnsc_zwallet_update_74  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | feature/update_74  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**